example/glue/ObjectModel.hpp

```diff
@@ -97,425 +59,207 @@ class GC_ObjectModel
-		while (oldFlags != MM_AtomicOperations::lockCompareExchange(flagsPtr, (uintptr_t)oldFlags, (uintptr_t)newFlags));
-#endif /* defined(OMR_INTERP_COMPRESSED_OBJECT_HEADER) */
-
-		return result;
+		fomrobject_t *headerSlotAddress = getObjectHeaderSlotAddress(objectPtr);
+		return *headerSlotAddress & (((fomrobject_t)1 << _objectMetadataSlotFlagsShift) - 1);
 	}
 
 	/**
-	 * Extract the flag bits from an unforwarded object.
+	 * Languages that wish to take full advantage of OMR's generational heap may elect not to
+	 * include some object components in initial allocations of objects. For example, components
+	 * that are meaningful only for tenured objects may be excluded from initial allocation in
+	 * the nursery and added only if/when the object is promoted to tenure space. In that case,
+	 * this method would determine whether or not the object had been promoted and return the
+	 * actual size on that basis.
 	 *
-	 * This method will assert if the object has been marked as forwarded.
+	 * Returned value must include size of object header and any padding bytes added for heap
+	 * alignment.
 	 *
-	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the object
-	 * @return the flag bits from the object encapsulated by forwardedHeader
-	 * @see MM_ForwardingHeader::isForwardedObject()
+	 * @param[in] objectPtr points to the object to obtain size for
+	 * @return actual total size of the object, in bytes, as of time of call
 	 */
 	MMINLINE uintptr_t
-	getPreservedFlags(MM_ForwardedHeader *forwardedHeader)
+	getConsumedSizeInBytesWithHeader(omrobjectptr_t objectPtr)
 	{
-		return (uintptr_t)(forwardedHeader->getPreservedSlot()) & OMR_OBJECT_METADATA_FLAGS_MASK;
+		return getSizeInBytesWithHeader(objectPtr);
 	}
 
 	/**
-	 * Extract the age bits from an unforwarded object.
-	 *
-	 * This method will assert if the object has been marked as forwarded.
+	 * This method simply converts the consumed size of an object from bytes to (fomrobject_t)
+	 * slots. This implementation should be sufficient for all languages.
 	 *
-	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the object
-	 * @return the age bits from the object encapsulated by forwardedHeader
-	 * @see MM_ForwardingHeader::isForwardedObject()
+	 * @param[in] objectPtr points to the object to obtain size for
+	 * @return actual total size of the object, in slots, as of time of call
 	 */
 	MMINLINE uintptr_t
-	getPreservedAge(MM_ForwardedHeader *forwardedHeader)
+	getConsumedSizeInSlotsWithHeader(omrobjectptr_t objectPtr)
 	{
-		return ((uintptr_t)(forwardedHeader->getPreservedSlot()) & OMR_OBJECT_METADATA_AGE_MASK) >> OMR_OBJECT_METADATA_AGE_SHIFT;
+		return MM_Bits::convertBytesToSlots(getConsumedSizeInBytesWithHeader(objectPtr));
 	}
 
 	/**
-	 * Extract the size from an unforwarded object.
+	 * See comment for getConsumedSizeInBytesWithHeader(), above. This method would be called
```

Comment:
Not sure what needs to be updated in comment, which states that this function is called "just before the object is to be promoted into tenure space". So, in generational context, without specific reference to Scavenger.